plutus-core/executables/pir/Main.hs

```diff
@@ -1,106 +1,288 @@
-import Data.List (sortOn)
-import Data.Text qualified as T
-import GHC.Generics
-import Options.Applicative
 import PlutusCore qualified as PLC
+import PlutusCore.Compiler qualified as PLC
 import PlutusCore.Error (ParserErrorBundle (..))
 import PlutusCore.Executable.Common hiding (runPrint)
 import PlutusCore.Executable.Parsers
 import PlutusCore.Quote (runQuoteT)
 import PlutusIR as PIR
 import PlutusIR.Analysis.RetainedSize qualified as PIR
 import PlutusIR.Compiler qualified as PIR
+import PlutusIR.Core.Instance.Pretty ()
 import PlutusIR.Core.Plated
 import PlutusPrelude
+import UntypedPlutusCore qualified as UPLC
+
+import Control.Lens (coerced, (^..))
+import Control.Monad.Trans.Except (runExcept)
+import Control.Monad.Trans.Reader (runReader, runReaderT)
+import Data.ByteString.Lazy.Char8 qualified as BSL
+import Data.Csv qualified as Csv
+import Data.IntMap qualified as IM
+import Data.List (sortOn)
+import Data.Text qualified as T
+import Options.Applicative
 import Text.Megaparsec (errorBundlePretty)
 
-type PLCTerm  = PLC.Term PLC.TyName PLC.Name PLC.DefaultUni PLC.DefaultFun (PIR.Provenance ())
-type PIRError = PIR.Error PLC.DefaultUni PLC.DefaultFun (PIR.Provenance ())
-type PIRCompilationCtx a = PIR.CompilationCtx PLC.DefaultUni PLC.DefaultFun a
-data Command = Analyse IOSpec
-             | Compile COpts
-             | Print PrintOptions
-
-data COpts = COpts
-  { cIn       :: Input
-  , cOptimize :: Bool
-  }
-
-pCOpts :: Parser COpts
-pCOpts = COpts
-            <$> input
-            <*> switch' ( long "dont-optimize"
-                        <> help "Don't optimize"
-                        )
-  where
-    switch' :: Mod FlagFields Bool -> Parser Bool
-    switch' = fmap not . switch
-
-pPirOpts :: Parser Command
-pPirOpts = hsubparser $
-    command "analyse"
-        (info (Analyse <$> ioSpec) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise and analyse the program, " <>
-              "looking for variables with the largest retained size.")
-  <> command "compile"
-        (info (Compile <$> pCOpts) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise it, " <>
-              "and test if it can be successfully compiled to PLC.")
-  <> command "print"
-        (info (Print <$> printOpts) $
-            progDesc $
-              "Given a PIR program in flat format, " <>
-              "deserialise it and print it out textually.")
-
--- | Load flat pir and deserialize it
-loadPir :: Input -> IO (PirProg ())
-loadPir = loadASTfromFlat Named
-
-compile :: COpts -> PirProg () -> Either PIRError PLCTerm
-compile opts (PIR.Program _ pirT) = do
+
+type PIRErrorWithProvenance = PIR.Error PLC.DefaultUni PLC.DefaultFun (PIR.Provenance ())
+
+---------------- Types for command line options ----------------
+
+-- | A specialised format type for PIR. We don't support deBruijn or named deBruijn for PIR.
+data PirFormat = TextualPir | FlatNamed
```

Comment:
We only have flat with names for PIR: unlike PLC/UPLC, de Bruijn indices aren't allowed.  This means that I've had to implement special cases for some of the parsers and option types.  It might have been easier just to add suitable `Flat` instances so you can do exactly the same with PIR as you can with the other languages.

Reply:
I'll see if I can fix that.